### PR TITLE
User Referenced Payload

### PR DESCRIPTION
**Pull request by zahnno**
User referenced payload class with wizardlet listeners

### CODE REVIEW

File: src/foam/nanos/crunch/UserReferencedPayload.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
+/**
+ * @license
+ * Copyright 2021 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+ foam.CLASS({
+  package: 'foam.nanos.crunch',
+  name: 'UserReferencedPayload',
+  documentation: 'Capability user related payload class - attaches listeners to wizardlets if defined',
```
Comment (kgrgreer) [inline src/foam/nanos/crunch/UserReferencedPayload.js]: Could you please expand on the documentation. Maybe providing more context. Thx

File: src/foam/nanos/crunch/UserReferencedPayload.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      `
+    }
+  ],
+
+  methods: [
+    function init() {
+      // Initialize listeners on defined wizardlets in reference to their capability ids
+      if ( this.wizardlets ) {
+        this.wizardletsToListen.forEach(capabilityId => {
+          wizardlet = this.wizardlets.find(wizardlet => wizardlet.capability.id === capabilityId );
```
Comment (Nauna) [inline src/foam/nanos/crunch/UserReferencedPayload.js:39]: didn't know we can do a find this way 👍

**Review by Nauna — COMMENTED** [review]
So maybe I'm wrong - I agree with Kgr I'd want more documentation.

However, this seems like a direct duplicate of the usage of daoKey and contextDAOFindKey properties on the Capability model.
https://github.com/kgrgreer/foam3/blob/1d1c6c22485ea5fd94570eff4de83f67028d97c5/src/foam/nanos/crunch/Capability.js#L200-L212

If you search the property name in the capabilities.jrl you'll see them ... side note: wonder why we didn't use this for the bankAccount capabilities .... 

So that feature is defined https://github.com/kgrgreer/foam3/blob/1d1c6c22485ea5fd94570eff4de83f67028d97c5/src/foam/nanos/crunch/UserCapabilityJunctionRefine.js#L275-L279

which says 'given the capability.of(payload/data), the daoKey and the contextDAOFindKey, I will copy the payload onto the imported object from context and save it with a put into the daoKey.'

Thus if you want to save something onto another object, like the copyToUser() method here, just ensure the object is in the correct context - which subject is - and name the properties you want to copy from capability.of model the same as on the object model and presto ....

really don't feel any more complexity to the wizard is a good idea. 
Nonetheless Good implementation and interesting solution.

File: src/foam/nanos/crunch/UserReferencedPayload.js (hunk 2 above)
Comment (tharmaman) [inline src/foam/nanos/crunch/UserReferencedPayload.js:39]: In the case of MinMax, if choice A has prereq C and choice B has prereq C. There will be two different wizardlets instantiated which use prereq C so not sure if you tested this on a minmax might have some unexpected behaviour. But from what I read this is an opt-in feature correct?

**Review by tharmaman — CHANGES_REQUESTED** [review]

File: src/foam/nanos/crunch/UserReferencedPayload.js (hunk 2 above)
Comment (zahnno) [reply src/foam/nanos/crunch/UserReferencedPayload.js:39]: I'll have to take some time to review this. From what I was aware of duplicate prerequisites are combined and a unique prerequisites list is displayed in the wizard.

Comment (zahnno) [reply src/foam/nanos/crunch/UserReferencedPayload.js:39]: Duplicate prerequisites of minmax capabilities aren't combined... oddly enough - should be following suit to how prerequisites of non minmax capabilities are unique when presented in the wizard.

Anyways, the intended behaviour of applying a listener on the wizardlet is as expected. When a wizardlet with a sub listener changes, send a pub to any payload listening to those changes. In the case of the minmax duplicates, the payloads are listening to the change events.

File: src/foam/nanos/crunch/UserReferencedPayload.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+ */
+
+ foam.CLASS({
+  package: 'foam.nanos.crunch',
+  name: 'UserReferencedPayload',
+  documentation: `
+    Capability user related payload class.
+    Attaches listeners to wizardlets if defined.
+    Useful for extending payload classes that would potentially be used by
+    capable objects requiring user specific logic such as copying user values
```
Comment (tharmaman) [inline src/foam/nanos/crunch/UserReferencedPayload.js:14]: so capable objects here you mean both ucj and capable, or just capable (where the payload is stored on the object)?

Comment (zahnno) [reply src/foam/nanos/crunch/UserReferencedPayload.js:14]: Just capable objects

**Pull request closed without merging**